src/Symfony/Component/Security/Http/EventListener/VerifyAuthenticatorCredentialsListener.php

```diff
@@ -0,0 +1,69 @@
+<?php
+
+namespace Symfony\Component\Security\Http\EventListener;
+
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\Security\Core\Encoder\EncoderFactoryInterface;
+use Symfony\Component\Security\Core\Exception\BadCredentialsException;
+use Symfony\Component\Security\Http\Authenticator\Passport\Credentials\CustomCredentials;
+use Symfony\Component\Security\Http\Authenticator\Passport\Credentials\PasswordCredentials;
+use Symfony\Component\Security\Http\Event\VerifyAuthenticatorCredentialsEvent;
+
+/**
+ * This listeners uses the interfaces of authenticators to
+ * determine how to check credentials.
+ *
+ * @author Wouter de Jong <wouter@driveamber.com>
+ *
+ * @final
+ * @experimental in 5.1
+ */
+class VerifyAuthenticatorCredentialsListener implements EventSubscriberInterface
+{
+    private $encoderFactory;
+
+    public function __construct(EncoderFactoryInterface $encoderFactory)
+    {
+        $this->encoderFactory = $encoderFactory;
+    }
+
+    public function onAuthenticating(VerifyAuthenticatorCredentialsEvent $event): void
+    {
+        $passport = $event->getPassport();
+        if ($passport->hasBadge(PasswordCredentials::class)) {
+            // Use the password encoder to validate the credentials
+            $user = $passport->getUser();
```

Comment:
There is no `getUser()` method unless we check for the `UserPassportInterface`. If the `PasswordCredentials` is added but we do *not* have a `UserPassportInterface`, I think we should fail hard with an exception - that's a misconfiguration.

Reply:
Btw, if we started only dispatching the `LoginSuccessEvent` when there is a `UserPassportInterface`, it might help the "anonymous" problem and the code in the listeners. I'm still "thinking out loud" about if that event should be dispatched for anonymous or not...

Reply:
👍 for adding this exception in this class. That allows to create a custom password verification listener for non-user passwords (which is perfectly possible, but not when using `UserPasswordEncoder` as we do here)